### PR TITLE
CT-3740 Add mynewt sensitive attribute

### PR DESCRIPTION
**Pull request by t3zeng**

### CODE REVIEW

File: hw/mcu/nordic/nrf5340/nrf5340.ld [hunk 1]
```diff
@@ -152,6 +152,10 @@ SECTIONS
         *(.data*)
 
         *(.jcr)
+
+        __MYNEWT_SENSITIVE_START = .;
+        *(.mynewt_sensitive*)
+        __MYNEWT_SENSITIVE_END = .;
```
Comment (vrahane) [inline hw/mcu/nordic/nrf5340/nrf5340.ld]: So, I think these should be consistent with the other variables, so, can you please make it lowercase.

File: util/memmask/src/memmask.c [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+extern int __MYNEWT_SENSITIVE_START, __MYNEWT_SENSITIVE_END;
+
+/**
+ * Clear out sensitive memory area
+ */
+void
+memmask_clear_sensitive_area(void)
+{
+    memset(&__MYNEWT_SENSITIVE_START, 0xff, (&__MYNEWT_SENSITIVE_END-&__MYNEWT_SENSITIVE_START));
```
Comment (vrahane) [inline util/memmask/src/memmask.c]: Let's fill it up with a custom pattern and make it configurable. Change the pattern to be something like `0xbaadbaad` or something like that by default so, we can recognize it very easily in coredumps.

File: kernel/os/src/arch/cortex_m33/os_fault.c [hunk 3]
```diff
@@ -213,6 +215,8 @@ os_default_irq(struct trap_frame *tf)
     uint32_t orig_sp;
 #endif
 
+    memmask_clear_sensitive_area();
```
Comment (vrahane) [inline kernel/os/src/arch/cortex_m33/os_fault.c]: similar to `log_reboot` this should be made conditional.

File: util/memmask/include/memmask/memmask.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef H_UTIL_MEMMASK_
+#define H_UTIL_MEMMASK_
```
Comment (vrahane) [inline util/memmask/include/memmask/memmask.h]: I think the name should be something different. Maybe `memprotect`. That way it is obvious that we will clear out the protected section. Also, let's call the section `.mynewt_protected`.

File: kernel/os/src/arch/cortex_m4/os_fault.c [hunk 5]
```diff
@@ -168,6 +170,8 @@ os_default_irq(struct trap_frame *tf)
     uint32_t orig_sp;
 #endif
 
+    memmask_clear_sensitive_area();
```
Comment (vrahane) [inline kernel/os/src/arch/cortex_m4/os_fault.c]: Let's make this conditional like `log_reboot`

File: hw/mcu/nordic/nrf52xxx/nrf52.ld [hunk 6]
```diff
@@ -145,6 +145,10 @@ SECTIONS
         *(.data*)
 
         *(.jcr)
+
+        __MYNEWT_SENSITIVE_START = .;
+        *(.mynewt_sensitive*)
+        __MYNEWT_SENSITIVE_END = .;
```
Comment (vrahane) [inline hw/mcu/nordic/nrf52xxx/nrf52.ld]: So, I think these should be consistent with the other variables, so, can you please make it lowercase.

File: hw/mcu/nordic/nrf5340/nrf5340.ld (hunk 1 above)
Comment (t3zeng) [reply hw/mcu/nordic/nrf5340/nrf5340.ld]: changed these to lower case

File: kernel/os/src/arch/cortex_m33/os_fault.c (hunk 3 above)
Comment (t3zeng) [reply kernel/os/src/arch/cortex_m33/os_fault.c]: Added a syscfg condition to enable the clearing feature

File: util/memmask/include/memmask/memmask.h (hunk 4 above)
Comment (t3zeng) [reply util/memmask/include/memmask/memmask.h]: I renamed everything from memmask to memprotect

File: kernel/os/src/arch/cortex_m4/os_fault.c (hunk 5 above)
Comment (t3zeng) [reply kernel/os/src/arch/cortex_m4/os_fault.c]: Done

File: hw/mcu/nordic/nrf52xxx/nrf52.ld (hunk 6 above)
Comment (t3zeng) [reply hw/mcu/nordic/nrf52xxx/nrf52.ld]: lower cased

File: util/memmask/src/memmask.c (hunk 2 above)
Comment (t3zeng) [reply util/memmask/src/memmask.c]: For a 4 byte pattern like `0xbaadbaad` I think we'll have to replace the memset with a 4 byte pointer that loops through each block from `__MYNEWT_SENSITIVE_START` to `__MYNEWT_SENSITIVE_END` and enforce sizing of the custom pattern

**Pull request closed without merging**